src/pe/debug.rs

```diff
@@ -201,61 +463,660 @@ pub struct CodeviewPDB20DebugInfo<'a> {
+///
+/// PGO is a compiler optimization technique that uses data collected from program
+/// execution to optimize code layout and improve runtime performance. This structure
+/// contains details such as the relative virtual address (RVA), size, and the associated
+/// name of the function or data block for which PGO data is provided.
+///
+/// <https://devblogs.microsoft.com/cppblog/pogo>
+#[repr(C)]
+#[derive(Debug, PartialEq, Copy, Clone, Default)]
+pub struct POGOInfo<'a> {
+    /// The signature of POGO debug directory entry is always first 4-bytes
+    ///
+    /// Either one of:
+    ///
+    /// - [`IMAGE_DEBUG_POGO_SIGNATURE_LTCG`]
+    /// - [`IMAGE_DEBUG_POGO_SIGNATURE_PGU`]
+    pub signature: u32,
+    /// Raw bytes of POGO debug entry, without first 4-bytes signatures field
+    pub data: &'a [u8],
+}
+
+/// Represents the `IMAGE_DEBUG_POGO_ENTRY` structure
+#[derive(Debug, PartialEq, Copy, Clone, Default)]
+pub struct POGOInfoEntry<'a> {
+    /// The relative virtual address (RVA) of the PGO data.
+    pub rva: u32,
+    /// The size of the PGO data block.
+    pub size: u32,
+    /// The name of the function or data block associated with the PGO data as a byte slice.
+    ///
+    /// This may contain a null-terminated string that represents the function or block
+    /// name for which PGO optimization data is provided.
+    pub name: &'a [u8],
+}
+
+/// Iterator over POGO entries in [`POGOInfo`].
+#[derive(Debug, Copy, Clone)]
+pub struct POGOEntryIterator<'a> {
+    /// The raw data of [`POGOInfo::data`] without the signature field
+    pub data: &'a [u8],
+}
+
+/// Indicates the PGO signature for Link-Time Code Generation (LTCG) (`u32` hex representation of `'LTCG'`).
+///
+/// This constant is used in the `IMAGE_DEBUG_DIRECTORY` to identify sections of
+/// PGO data generated specifically for LTCG optimizations.
+pub const IMAGE_DEBUG_POGO_SIGNATURE_LTCG: u32 = 0x4C544347;
+/// Indicates the PGO signature for Profile-Guided Optimization (PGO) updates (PGU) (`u32` hex representation of `'PGU\0'`).
+///
+/// This constant is used to signify sections of PGO data associated with PGO updates,
+/// which are incremental optimizations based on profiling data collected over time.
+pub const IMAGE_DEBUG_POGO_SIGNATURE_PGU: u32 = 0x50475500;
+/// Size of [`IMAGE_DEBUG_POGO_SIGNATURE_LTCG`] or [`IMAGE_DEBUG_POGO_SIGNATURE_PGU`]
+pub const POGO_SIGNATURE_SIZE: usize = core::mem::size_of::<u32>();
+
+impl<'a> POGOInfo<'a> {
+    pub fn parse(
+        bytes: &'a [u8],
+        idd: ImageDebugDirectoryIterator<'_>,
+    ) -> error::Result<Option<Self>> {
+        Self::parse_with_opts(bytes, idd, &options::ParseOptions::default())
+    }
+
+    pub fn parse_with_opts(
+        bytes: &'a [u8],
+        idd: ImageDebugDirectoryIterator<'_>,
+        opts: &options::ParseOptions,
+    ) -> error::Result<Option<Self>> {
+        let idd = idd.collect::<Result<Vec<_>, _>>()?;
```

Comment:
ditto allocation; it seems like all of these allocate simply to call `find` on the type iiuc?